rfcs/0052-dynamic-ids.md

```diff
@@ -0,0 +1,112 @@
+---
+feature: dynamic-ids
+start-date: 2019-09-05
+author: Silvan Mosberger
+co-authors: (find a buddy later to help our with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: https://github.com/NixOS/nixpkgs/pull/65698
+---
+
+# Summary
+[summary]: #summary
+
+A lot of NixOS modules are [assigning static uids/gids](https://github.com/NixOS/nixpkgs/blob/044cc701c23ede96355eb1dc997985b4dfac0372/nixos/modules/misc/ids.nix#L36) to their users. This has resulted in less than 90 static ids left in the reserved range from 0 to 400.
+
+This RFC deprecates the practice of doing that and instead suggests to
+- If applicable use systemd's `DynamicUser`
+- Otherwise let NixOS assign dynamic persistent ids, which happens automatically when `users.users.<name?>.uid`/`users.groups.<name?>.gid` is not set. For users, `users.users.<name?>.isSystemUser` should be set so that only uids under 1000 are used by NixOS services.
+
+Only in special circumstances are static ids allowed anymore.
```

Comment:
I think it would be a good idea to list such special circumstances. I would guess that something that doesn't fit `StateDir` model and locks one out of the system on failure (at least from remote access) would be an example?

Reply:
Not sure if there's any program that would actually do that. As an actual example, I think there is NFS3 which might not be able to map ids from other machines to local ones, but NFS4 doesn't have that problem anymore I'm pretty sure.

In general the problem might be programs that transfer files between machines without adjusting the ids accordingly. And arguably since ids don't have to be the same on different machines, that's really a bug in the program then.

Another example might be programs that mention that their user should have a certain fixed uid, because it might be hardcoded in the source, or because it might not have code to handle different ones on different machines. No concrete I know, but I wouldn't be surprised to find this somewhere.

Reply:
Or one could reword it to: "usage of static ids has to be explicitly justified", so as to put the burden on the code's author rather than the RFC's (assuming we could even come up with an exhaustive list of reasons where static IDs are necessary)

Reply:
Well, «explicitly justified» is not really a clear description of _level_ of burden of proof.